src/middlewared/middlewared/etc_files/smb_configure.py

```diff
@@ -0,0 +1,352 @@
+        os.chmod(TMP_PRIVATEDIR, 0o700)
+
+    with open(TMP_SMBPASSWD, "w") as f:
+        for user in conf['smb_users']:
+            f.write(user['bsdusr_smbhash'] + '\n')
+
+    return True
+
+async def disable_passdb_users(smb_users):
+    for user in smb_users:
+        if user['bsdusr_locked'] or user['bsdusr_password_disabled']:
+            flags = '-d'
+        else:
+            flags = '-e'
+
+        set_pdb_flags = await run([SMBPASSWDCMD, flags, user["bsdusr_username"]], check=False)
+        if set_pdb_flags.returncode != 0:
+            logger.debug(f'failed to set flags {flags} for user {user["bsdusr_username"]} in passdb.tdb with error: {set_pdb_flags.stderr.decode()}')
+            return False
+
+    return True
+
+async def import_local_users(middleware, conf):
+    passdb_file = f"{conf['privatedir']}/passdb.tdb"
+    passdb_usercount = await count_passdb_users()
+    if len(conf['smb_users']) == passdb_usercount[1]:
+        logger.debug(f'User count in config file and passdb is identical. Bypassing import')
+        return True
+
+    ret = await write_legacy_smbpasswd(middleware, conf)
+
+    if os.path.exists(passdb_file):
+        logger.debug('Removing out-of-sync passdb.tdb file')
+        os.unlink(passdb_file)
+
+    pdb_import = await run([PDBCMD, '-d 1',
+                           f"-i smbpasswd:{TMP_SMBPASSWD}"],
+                           check=False)
+
+    os.unlink(TMP_SMBPASSWD)
+    if pdb_import.returncode != 0:
+        logger.debug(f'Import of temp smbpasswd file failed with error: {pdb_import.stderr.decode()}')
+        return False
+
+    ret = await disable_passdb_users(conf['smb_users'])
+    if not ret:
+        return False
+
+"""
+    Validate contents of group_mapping.tdb, which maps local Unix groups to Windows group
+    SIDs. This file should:
+    1) contain no duplicate or inconsistent entries
+    2) contain no group names that are identical to usernames
+
+    To do: add validation of SID values in the tdb file.
+"""
+async def groupmap_add(config, unixgroup, ntgroup):
+    gm_add = await run([NETCMD, '-d 1', 'groupmap', 'add',
+                       f'unixgroup={unixgroup}',
+                       f'ntgroup={ntgroup}'],
+                       check=False)
+
+    if gm_add.returncode != 0:
+        logger.debug(f'Failed to map {unixgroup} to nt group {ntgroup}: {gm_add.stderr.decode()}')
+        return False
+
+    return True
+
+async def is_disallowed_group(middleware, conf, groupmap, group):
+    disallowed_list = ['USERS', 'ADMINISTRATORS']
+    for user in conf['smb_users']:
+        disallowed_list.append(user['bsdusr_username'].upper())
+    for gm in groupmap:
+        disallowed_list.append(gm['unixgroup'].upper())
+
+    if group.upper() in disallowed_list:
+        return True
+
+    return False
+
+async def get_groups(middleware):
+    _groups = {}
+    groups = await middleware.call('datastore.query', 'account.bsdGroups', [('bsdgrp_builtin', '=', False)])
```

Comment:
Shall we use group.query?

A second query could be made with
`user.query, [["id", "in", group["users"]]]`